GH-10083: Apply Nullability to `file` module

artembilan opened the pull request:
Related to: https://github.com/spring-projects/spring-integration/issues/10083

* The change in the `file.dsl` package has affected the `ComponentsRegistration` contract in the core module. And that one has led to respective fixes in the `AMQP`, `JPA`, `Kafka` modules for DSL packages
* Fix `SimpleJsonSerializer` contract to return an empty JSON instead of `null`. Something similar to what Jackson 3 provides for us by default.
* Rework a `WatchServiceDirectoryScanner` logic to initialize its internals in the ctor to avoid null-away errors
* Fix `FileWritingMessageHandler`, `HeadDirectoryScanner`, and `RecursiveDirectoryScanner` code style and logic according to IDE suggestions and new Nullability functionality
* Fix classes in the `file.config` package according to new Nullability logic
* Implement Nullability for `FileListFilter` abstraction and its implementations
* Fix `FileChannelCache` for a proper constant name.
* Respectively, fix the `NioFileLockerTests` to use a new `FileChannelCache.CHANNEL_CACHE` constant
* The remote directory could be null for some protocol operations. Therefore implement that respectively in the `FileInfo`, `Session` and `RemoteFileOperations`

The change might look big, but it is inevitable and pretty straightforward

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-10083: Apply Nullability to `file` module

Related to: https://github.com/spring-projects/spring-integration/issues/10083

* The change in the `file.dsl` package has affected the `ComponentsRegistration` contract in the core module.
And that one has led to respective fixes in the `AMQP`, `JPA`, `Kafka` modules for DSL packages
* Fix `SimpleJsonSerializer` contract to return an empty JSON instead of `null`.
Something similar to what Jackson 3 provides for us by default.
* Rework a `WatchServiceDirectoryScanner` logic to initialize its internals in the ctor to avoid null-away errors
* Fix `FileWritingMessageHandler`, `HeadDirectoryScanner`, and `RecursiveDirectoryScanner` code style and logic according to IDE suggestions and new Nullability functionality
* Fix classes in the `file.config` package according to new Nullability logic
* Implement Nullability for `FileListFilter` abstraction and its implementations
* Fix `FileChannelCache` for a proper constant name.
* Respectively, fix the `NioFileLockerTests` to use a new `FileChannelCache.CHANNEL_CACHE` constant
* The remote directory could be null for some protocol operations.
Therefore implement that respectively in the `FileInfo`, `Session` and `RemoteFileOperations`

The change might look big, but it is inevitable and pretty straightforward

ngocnhan-tran1996 commented on spring-integration-file/src/main/java/org/springframework/integration/file/FileReadingMessageSource.java:
```diff
@@ -370,7 +372,7 @@ protected void onInit() {
 
 	@Override
-	protected AbstractIntegrationMessageBuilder<File> doReceive() {
+	protected @org.jspecify.annotations.Nullable AbstractIntegrationMessageBuilder<File> doReceive() {
```
I think `@Nullable` is enough, no need for the fully qualified class name.

artembilan committed:
* Remove redundant FQCN for `@Nullable` on the `FileReadingMessageSource.doReceive()`.
That was a first JSpecify annotation changed in this class without replacing existing import for deprecated one from SF
* Fix `FileTransferringMessageHandler.handleMessageInternal()` logic to fail when file could not be transferred to remote dir.
Previously it returned silently.
The error was only logged in the `RemoteFileTemplate`

cppwfs commented on spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java:
```diff
@@ -514,7 +514,7 @@ private void validateDestinationDirectory(File destinationDirectory, boolean aut
 
 	@Override // NOSONAR
-	protected Object handleRequestMessage(Message<?> requestMessage) {
+	protected @Nullable Object handleRequestMessage(Message<?> requestMessage) {
```
Do we want to remove the `// NO SONAR`?

cppwfs commented on spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java:
```diff
@@ -967,11 +969,13 @@ public void flushIfNeeded(FlushPredicate flushPredicate) {
-	private Map<String, FileState> findFilesToFlush(MessageFlushPredicate flushPredicate, Message<?> filterMessage) {
+	private Map<String, FileState> findFilesToFlush(MessageFlushPredicate flushPredicate,
+			@Nullable Message<?> filterMessage) {
+
```
Remove the extra space.

artembilan replied:
No. That's my preference: if method head is multi-line, then we need extra blank line before method body.
It makes the code much easier to read.

cppwfs commented on spring-integration-file/src/main/java/org/springframework/integration/file/config/FileReadingMessageSourceFactoryBean.java:
```diff
@@ -128,20 +132,19 @@ protected FileReadingMessageSource createInstance() {
 		else {
 			this.source = new FileReadingMessageSource();
 		}
+		Assert.notNull(this.directory, "The 'directory' must be provided.");
```
This can be removed because the setDirectory already checks for null.

artembilan replied:
Right. And what if `setDirectory()` was not called?

cppwfs commented on spring-integration-file/src/main/java/org/springframework/integration/file/DefaultDirectoryScanner.java:
```diff
@@ -82,17 +84,19 @@ protected FileLocker getLocker() {
 	@Override
 	public boolean tryClaim(File file) {
-		return (this.locker == null) || this.locker.lock(file);
+		FileLocker lockerToUse = this.locker;
```
Was this change made for readability?

cppwfs reviewed (changes requested):
Looks good.   Just some nitpicks

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/tail/FileTailingMessageProducerSupport.java:
```diff
@@ -169,20 +166,16 @@ protected void doStart() {
-		if (this.idleEventScheduledFuture != null) {
-			this.idleEventScheduledFuture.cancel(true);
+		ScheduledFuture<?> idleEventScheduledFutureToCancel = this.idleEventScheduledFuture;
+		if (idleEventScheduledFutureToCancel != null) {
```
Same NPE race condition as in many other places.
If we don't change it, then IDE complains that `this.idleEventScheduledFuture.cancel(true);` could fail with NPE.
And I am kinda agreed with that: there is no guarantee at runtime that `this.idleEventScheduledFuture` does not become `null` after we enter that `if` block.

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/aop/RotationPolicy.java:
```diff
@@ -54,33 +53,29 @@ public interface RotationPolicy {
 			return this.directory;
 		}
 
-		@Override
```
True. Because Java `record` does that for us automatically. From Google: 
```
Records use a streamlined syntax that automatically generates common methods, such as:

    A canonical constructor that takes all record components as arguments.
    Accessor methods for each component (e.g., name() for a name component).
    equals(), hashCode(), and toString() methods based on the record's components.
```
More info in: https://www.baeldung.com/java-record-keyword

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizingMessageSource.java:
```diff
@@ -256,11 +259,11 @@ public boolean isRunning() {
 	 * If the result is null, it attempts to sync up with the remote directory to populate the file source.
 	 * At most, maxFetchSize files will be fetched.
-	 * Then, it polls the file source again and returns the result, whether or not it is null.
+	 * Then, it polls the file source again and returns the result, whether it is null.
```
Sure!
But if this is only concern in this review, I wouldn't bother 😉

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/FileInfo.java:
```diff
@@ -55,12 +57,14 @@ public interface FileInfo<F> {
 	 * @return a string representing the permissions of the remote
 	 * file (e.g. -rw-r--r--).
 	 */
+	@Nullable
```
Not on methods in the interfaces.

artembilan committed:
* Fix language in `AbstractInboundFileSynchronizingMessageSource.doReceive()` JavaDoc
* Bring back a proper JavaDoc for `RemoteFileOperations.rename()`
* Extract pattern variable for the `if` in the `FileWritingMessageHandler.handleRequestMessage()`
* Fix indents in the `HeadDirectoryScanner.HeadFilter` record

cppwfs reviewed (approved):
LGTM
